### DISCUSSION TITLE
Started on graphic interface for Lima using Qt

### DISCUSSION THREAD

**Discussion by afbjorklund** [lima-vm/lima#395] [Ideas]
Similar to the PoC that I did for Podman:
https://github.com/afbjorklund/lima-gui

Added some design docs to the wiki of the project:
https://github.com/afbjorklund/lima-gui/wiki/Main-list

* Cross-platform (Linux, Mac, Win)

* System Tray Icon + Main Window

* Built-in terminal, with color support (thanks [QTermWidget](https://github.com/lxqt/qtermwidget))

* Built-in editor, with syntax highlighting (thanks [QSourceHighlite](https://github.com/Waqar144/QSourceHighlite))

Uses Qt 5, since that version comes standard with Ubuntu 20.04
Might be more hassle to set up on Mac, but I think it is in `brew` ?

Once it has more features implemented (currently it knows "shell"),
I would be happy to move it over to the "lima-vm" organization too.

**Comment by AkihiroSuda** [comment]
Thanks for working on this!

> Might be more hassle to set up on Mac, but I think it is in brew ?

qtermwidget seems missing in `brew`, so somebody will have to submit a Formula to Homebrew

> I would be happy to move it over to the "lima-vm" organization too.

Thanks, I'm happy to accept your repo, but if we move the project to some foundation like CNCF, we might have to move off non-Apache licensed repos. We are not discussing such plan (yet), though.

Also please see https://github.com/afbjorklund/lima-gui/issues/1 for the licensing concern.

**Comment by afbjorklund** [comment]
I can see how GPL is a concern, but BSD and LGPL are alright I hope

**Comment by afbjorklund** [comment]
Now using external $TERMINAL (x-terminal-emulator), so no GPL library.

**Comment by afbjorklund** [comment]
Now the create/start/stop/remove buttons work as well, plus a new refresh button...

The refresh button is normally not needed, but can be used to sync with e.g. CLI

The next thing is to add columns for the state of each instance, and not just the name.

This also enables disablings button, such as requiring stop before remove and so on.

**Comment by afbjorklund** [comment]
~Now using JSON list, but so far only status is showing.~

~Need to update the buttons, and add some more columns...~

Now also showing the arch, cpus, memory and disk columns

Rearranged the buttons a bit, and updating buttons on select

**Comment by afbjorklund** [comment]
I made a binary release, in case you don't want to build from source:

https://github.com/afbjorklund/lima-gui/releases/tag/v0.1.0

**Comment by afbjorklund** [comment]
One question that remains, is if there should be anything for `nerdctl` ?

Currently it mostly handles lima instances, but no containers/images...

**Comment by AkihiroSuda** [comment]
> One question that remains, is if there should be anything for `nerdctl` ?

Yes :+1:

**Comment by afbjorklund** [comment]
The thing is that lima(ctl) itself doesn't have any nerdctl-specific features, beyond `nerdctl.lima` and `containerd:` installer.

Beyond that, everything is run over the `shell` command, and thus delegated to the `nerdctl` command inside the VM ?

So one could argue that it would be up to the container runtime to present an interface, and not do it "inside" lima itself.

For instance the new Podman machine posts a link to [Cockpit](https://cockpit-project.org/), and for Docker one can deploy something like [Portainer](https://www.portainer.io/products/community-edition) (CE).

----

So far I haven't seen much except for Rancher Desktop that supports containerd, though. (no `docker.sock` available)

If you exclude Kubernetes and Nomad, and some CLI apps, that is. So there is not much available for "linking" to, either.

It wouldn't be _impossible_ to wrap `lima nerdctl` (or `lima sudo nerdctl`) the same way that we wrap `limactl` now.

But it would be some scope creep and additions, and definitely means it would be back to the [drawing board](https://github.com/afbjorklund/lima-gui/wiki) (Pencil Sketch)

----

### images
```console
$ lima nerdctl images
REPOSITORY    TAG       IMAGE ID        CREATED           SIZE
busybox       latest    f7ca5a32c10d    11 seconds ago    1.3 MiB
```

### containers
```console
$ lima nerdctl ps -a
CONTAINER ID    IMAGE                               COMMAND    CREATED           STATUS                       PORTS    NAMES
46d487c5b23c    docker.io/library/busybox:latest    "true"     23 seconds ago    Exited (0) 23 seconds ago          
```

It would be "nice" to support all three container runtimes, as well.
* docker
* podman
* nerdctl

Unfortunately there are some big holes in CRI, but they are "similar".

**Comment by afbjorklund** [comment]
The initial release is available here: https://github.com/afbjorklund/lima-gui/releases/tag/v0.1.0

Also fixed compatibility with Qt6, since the release (it is compatible with both versions of Qt)

It seems unlikely that it will get any container or image functionality in the near future, though.

At this point I consider it more or less feature complete, will see about the Windows version...
